src/Orleans/Utils/IKeyedServiceDescriptor.cs

```diff
@@ -0,0 +1,85 @@
+
+namespace Orleans
+{
+    public static class KeyedServiceDescriptorExtensions
+    {
+        /// <summary>
+        /// Add supported for keyed services to a service collection
+        /// </summary>
+        public static IServiceCollection SupportKeyedServices(this IServiceCollection collection)
+        {
+            return collection.AddSingleton(typeof(IKeyedServiceCollection<,>), typeof(KeyedServiceCollection<,>));
+        }
+
+        /// <summary>
+        /// Associates a key with a specialization of a service type.
+        /// </summary>
+        public static IServiceCollection WithKey<TKey, TService, TSpecialization>(this IServiceCollection collection, TKey key)
+            where TService : class
+            where TSpecialization : class, TService
+        {
+            return collection.AddSingleton<IKeyedServiceDescriptor<TKey, TService>>(sp => new KeyedServiceDescriptor<TKey, TService, TSpecialization>(key));
+        }
+
+        /// <summary>
+        /// Associates a name with a specialization of a service type.
+        /// </summary>
+        public static IServiceCollection WithName<TService, TSpecialization>(this IServiceCollection collection, string name)
+            where TService : class
+            where TSpecialization : class, TService
+        {
+            return collection.WithKey<string, TService, TSpecialization>(name);
+        }
+    }
+
+    internal interface IKeyedServiceDescriptor<TKey, TService> : IEquatable<TKey>
+        where TService : class
+    {
+        Type Service { get; }
+    }
+
+    internal class KeyedServiceDescriptor<TKey, TService, TInstance> : IKeyedServiceDescriptor<TKey, TService>
+            where TService : class
+            where TInstance : TService
+    {
+        private readonly TKey key;
+
+        public KeyedServiceDescriptor(TKey key)
+        {
+            this.key = key;
+            this.Service = typeof(TInstance);
+        }
+
+        public Type Service { get; }
+
+        public bool Equals(TKey other)
+        {
+            return Equals(key, other);
+        }
+    }
+
+    internal class KeyedServiceCollection<TKey, TService> : IKeyedServiceCollection<TKey, TService>
+        where TService : class
+    {
+        private readonly IKeyedServiceDescriptor<TKey, TService>[] keyedServices;
+        private readonly IServiceProvider services;
+
+        public KeyedServiceCollection(IServiceProvider services)
+        {
+            this.keyedServices = services.GetServices<IKeyedServiceDescriptor<TKey, TService>>().ToArray();
+            this.services = services;
+        }
+
+        public TService GetService(TKey key)
+        {
+            IKeyedServiceDescriptor<TKey, TService> keyedService = this.keyedServices.FirstOrDefault(s => s.Equals(key));
+            return keyedService == null
+                ? default(TService)
+                : services.GetService(keyedService.Service) as TService;
```

Comment:
In the above code, the users will always get null, as no azure storage provider was ever registered.  As for the problem of mapping to multiple instance of the same specialization, yes, out of the box, this is not supported.  This only disambiguates between different specializations of a base service type.

For storage providers, we'd planned to simply put the existing manager behind the IKeyedServiceCollection interface, and use the existing managers.

Question, do we need disambiguation of separate instances of a service type, rather than specializations of that type?  I was not of the impression that other DI containers (autofac, unity) supported this, am I mistaken?